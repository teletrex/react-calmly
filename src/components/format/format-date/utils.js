/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */



import moment from 'moment-timezone';

import { DATE_FORMAT } from './presets';

const shortenYear = format => format.replace(/Y+/g, val => val.slice(2));

const shortenDayMonth = format => format.replace(/D+|M+/g, val => val.slice(1));

export const getFormats = (locale, longDateFormat) => {
  const format = moment.localeData(locale).longDateFormat(longDateFormat);
  const reverseFormat = format
    .split('')
    .reverse()
    .join('');
  const shortYearFormat = shortenYear(format);
  const shortDayMonthFormat = shortenDayMonth(format);
  const shortDayMonthYearFormat = shortenYear(shortDayMonthFormat);

  return {
    format,
    reverseFormat,
    shortYearFormat,
    shortDayMonthFormat,
    shortDayMonthYearFormat,
    yearOnly: 'YYYY',
  };
};

const separatorDictironary = {
  Dot: '.',
  Comma: ',',
  ForwardSlash: '/',
  Hyphen: '-',
};

const trimSeparators = format => format.replace(/^[./\-,]+|[./\-,]+$/g, '');
const replaceSeparatorBy = (format, newSeparator) => format.replace(/[.,/-]/g, newSeparator);

export const getAllSeparatorsShortDateFormats = locale => {
  const formats = getFormats(locale, DATE_FORMAT.SHORT_DATE);

  return Object.entries(separatorDictironary).reduce(
    (allFormats, [separatorName, separator]) => {
      Object.entries(formats).forEach(([formatName, format]) => {
        // eslint-disable-next-line no-param-reassign
        allFormats[`${formatName}${separatorName}`] = replaceSeparatorBy(
          trimSeparators(format),
          separator
        );
      });
      return allFormats;
    },
    { ...formats }
  );
};
