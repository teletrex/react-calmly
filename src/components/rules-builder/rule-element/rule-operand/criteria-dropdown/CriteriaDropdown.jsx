/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */




import React, { useMemo, useContext } from 'react';
import PropTypes from 'prop-types';
import uniqueId from 'lodash/uniqueId';
import  settings from '../../../../../settings';
import noop from 'lodash/noop';
import {useTranslation} from '../../../../../translation';
import {  Dropdown } from '@carbon/react';
// import { Dropdown } from '../../../../new-dropdown';
import RulesBuilderContext from '../../../RulesBuilderContext';

const { prefix } = settings;
const defaultClassName = `${prefix}--rule-criteria`;

const getCriteriaData = (items, key, value) => {
  let foundCriteria;
  items.some(function iter(item) {
    if (item[key] === value) {
      foundCriteria = item;
      return true;
    }
    return Array.isArray(item.sublist) && item.sublist.some(iter);
  });
  return foundCriteria;
};

const CriteriaDropdown = ({ className, selectedCriteriaKey, label, onChange }) => {
  const { t } = useTranslation();

  const { criteriaDropdown: items = [], isLight, isReadOnly } = useContext(RulesBuilderContext);

  const id = () => `CriteriaDropdown_${uniqueId()}`;

  const selectedCriteria = useMemo(() => getCriteriaData(items, 'id', selectedCriteriaKey), [
    items,
    selectedCriteriaKey,
  ]);

  const handleChangeCriteria = ({ selectedItem }) => onChange(selectedItem.id);

  if (isReadOnly) {
    return (
      <span className={`${defaultClassName}--readonly`}>
        {selectedCriteriaKey ? selectedCriteria.text : t('Select criteria')}
      </span>
    );
  }

  return (
    <Dropdown
      key={`select_criteria_${selectedCriteria}`}
      ariaLabel={t('Select criteria')}
      className={className}
      id={`${id}-rule-criteria`}
      initialSelectedItem={selectedCriteria}
      isInline
      isLight={isLight}
      items={items}
      onChange={handleChangeCriteria}
      placeholder={label || t('Select criteria')}
    />
  );
};

CriteriaDropdown.propTypes = {
  className: PropTypes.string,
  label: PropTypes.string,
  onChange: PropTypes.func,
  selectedCriteriaKey: PropTypes.string,
};

CriteriaDropdown.defaultProps = {
  className: '',
  label: undefined,
  onChange: noop,
  selectedCriteriaKey: null,
};

export default CriteriaDropdown;
