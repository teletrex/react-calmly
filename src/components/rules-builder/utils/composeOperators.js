/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import compact from 'lodash/compact';
import map from 'lodash/map';
import isString from 'lodash/isString';
import find from 'lodash/find';

export const composeOperators = (operators, allowedOperators) =>
  compact(
    map(operators, operator => {
      if (isString(operator)) {
        return find(allowedOperators, ({ value }) => value === operator);
      }

      return operator;
    })
  );
