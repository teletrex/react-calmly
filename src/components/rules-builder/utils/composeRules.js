/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */




import isArray from 'lodash/isArray';
import omit from 'lodash/omit';
import uid from '@carbon/react/es/tools/uniqueId';

const mapRule = (rule, level, oneConjunctionMode) => {
  if (isArray(rule.data)) {
    let operands = 0;
    let maxLevelInside = 0;
    const data = rule.data.map(item => {
      const ret = mapRule(item, level + 1);
      operands += ret.operands;
      maxLevelInside = Math.max(maxLevelInside, ret.maxLevelInside);
      return ret;
    });
    return {
      ...rule,
      id: uid(),
      conjunction: oneConjunctionMode || rule.conjunction,
      level,
      maxLevelInside: maxLevelInside + 1,
      operands,
      data,
    };
  }

  return {
    ...rule,
    id: uid(),
    level,
    maxLevelInside: 0,
    operands: 1,
    data: rule.data,
  };
};

const mapRuleElements = (rule, createRule) => {
  const data = isArray(rule.data)
    ? rule.data.map(item => mapRuleElements(item, createRule))
    : createRule(rule.criteriaKey, rule.data, rule.timeframe);

  return {
    ...omit(rule, 'timeframe'),
    data,
  };
};

const composeRules = (rules, createRule, oneConjunctionMode) =>
  mapRuleElements(mapRule(rules, 0, oneConjunctionMode) || [], createRule);

export default composeRules;
