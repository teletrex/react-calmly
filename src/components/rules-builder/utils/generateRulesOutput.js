/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import filter from 'lodash/filter';
import isArray from 'lodash/isArray';
import isEmpty from 'lodash/isEmpty';
import map from 'lodash/map';
import omitBy from 'lodash/omitBy';
import pick from 'lodash/pick';

const generateRulesOutput = rules => {
  const { conjunction: topLevelConjunction, data: rulesData } = rules;

  const outputData = map(filter(rulesData, 'data'), rule => {
    const { data, conjunction, criteriaKey } = rule;
    const isGroup = isArray(data) && conjunction && !criteriaKey;

    return isGroup
      ? generateRulesOutput(rule)
      : omitBy(
          {
            criteriaKey,
            data: pick(data.ruleData, 'operator', 'value'),
            timeframe: pick(data.ruleTimeframe, 'operator', 'value'),
          },
          isEmpty
        );
  });

  return {
    conjunction: topLevelConjunction,
    data: outputData,
  };
};

export default generateRulesOutput;
