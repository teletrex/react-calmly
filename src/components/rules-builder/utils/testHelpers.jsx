/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';

import RulesBuilderContext from '../RulesBuilderContext';
import {
  ERROR_RELEGATION_REASON_TAG,
  relegateErrors,
} from '../../utils/test-helpers/relegate-errors';

export const datePickerErrorWrapper = fn => {
  relegateErrors(
    {
      reasonTag: ERROR_RELEGATION_REASON_TAG.FAILED_TO_RESOLVE,
      reason: "Couldn't trace the 'uncontrolled input error' in date component.",
      expectedErrorCount: 1,
    },
    () => {
      fn();
    }
  );
};

export const renderWithContext = (component, contextValue = {}) =>
  render(
    <RulesBuilderContext.Provider value={contextValue}>{component}</RulesBuilderContext.Provider>
  );
