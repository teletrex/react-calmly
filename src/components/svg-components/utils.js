/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const getSVGSizeProps = (size, height, width) => {
  const props = size ? { width: size, height: size } : {};
  if (height) {
    props.height = height;
  }
  if (width) {
    props.width = width;
  }
  return props;
};
