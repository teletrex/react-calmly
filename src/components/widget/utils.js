/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */




import { createCssClassBuilder } from '../../utils';
import { sortObjectByValue } from '../../utils/ObjectUtils';

import { breakpoints } from './constants';

export const widgetClassBuilder = createCssClassBuilder('widget');

export const getBreakpointFromWidth = width => {
  const sortedBreakpoints = sortObjectByValue(breakpoints);

  const lowerBP = sortedBreakpoints[0][0];

  return sortedBreakpoints.reduce((currentBP, [targetBP, BPWidth]) => {
    if (width > BPWidth) {
      return targetBP;
    }
    return currentBP;
  }, lowerBP);
};

export const getGetNewPosition = layout => {
  const xByRow = { 0: 0 };
  let lowestRow = 0;

  layout.forEach(item => {
    const nextAvailableXRelativeToItem = item.x + item.w;
    const lowestItemRow = item.y + item.h - 1;

    // the item can occupy more than one row
    // save his x to each of those rows
    for (let l = 0; l < item.h; l++) {
      const currentRow = l + item.y;
      const rowX = xByRow[currentRow] || 0;
      if (nextAvailableXRelativeToItem > rowX) {
        xByRow[currentRow] = nextAvailableXRelativeToItem;
      }
    }

    if (lowestItemRow > lowestRow) {
      lowestRow = lowestItemRow;
    }
  });

  const proposedX = xByRow[lowestRow];

  return {
    // since we have 6 columns and min width is 2,
    // we cannot place an element after the 4th column,
    // we do, instead, put it on 0 becase it will land on beggining of the next line.
    x: proposedX > 4 ? 0 : proposedX,
    // always put it one line below because even if we don't need to
    // break line, react-grid-layout will vertically compress.
    y: lowestRow + 1,
  };
};
