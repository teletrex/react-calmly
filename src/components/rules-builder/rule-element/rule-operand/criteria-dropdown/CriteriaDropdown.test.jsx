/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import userEvent from '@testing-library/user-event';

import { renderWithContext } from '../../../utils/testHelpers';

import CriteriaDropdown from './CriteriaDropdown';

const criteriaDropdown = [
  {
    id: 'foo-criteria',
    label: 'FooCriteria',
  },
];

const mockOnCriteriaChange = jest.fn();
const props = {
  onChange: mockOnCriteriaChange,
};

describe('CriteriaDropdown', () => {
  it('should render correctly', () => {
    const wrapper = renderWithContext(<CriteriaDropdown {...props} />, { criteriaDropdown });
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it('should render readOnly and behave like it', () => {
    const wrapper = renderWithContext(<CriteriaDropdown {...props} />, {
      criteriaDropdown,
      isReadOnly: true,
    });
    const readOnly = wrapper.container.querySelector('.bx--rule-criteria--readonly');

    expect(readOnly).not.toBe(null);
  });

  it('should render readOnly with selectedCriteriaKey', () => {
    const wrapper = renderWithContext(
      <CriteriaDropdown {...props} selectedCriteriaKey="foo-criteria" />,
      { criteriaDropdown, isReadOnly: true }
    );
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it('should properly click on criteria option', async () => {
    const { getByText } = renderWithContext(<CriteriaDropdown {...props} />, { criteriaDropdown });

    const criteriaDropdownComponent = getByText('Select criteria');
    userEvent.click(criteriaDropdownComponent);

    const criteria = getByText('FooCriteria');
    userEvent.click(criteria);

    expect(mockOnCriteriaChange).toHaveBeenCalledWith('foo-criteria');
  });
});
