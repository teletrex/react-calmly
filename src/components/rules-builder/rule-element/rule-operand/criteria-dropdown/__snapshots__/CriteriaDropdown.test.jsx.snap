// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CriteriaDropdown should render correctly 1`] = `
<DocumentFragment>
  <div
    class="bx-proposal-dropdown-wrapper bx-proposal-dropdown-wrapper--inline"
  >
    <div
      class="bx-proposal-dropdown"
      role="listbox"
    >
      <span>
        <div
          aria-label="Select criteria"
          class="bx-proposal-dropdown__field"
          role="button"
          tabindex="0"
        >
          <span
            class="bx-proposal-dropdown__field__label"
          >
            Select criteria
          </span>
          <div
            class="bx-proposal-dropdown__field__menu-icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              style="will-change: transform;"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
              />
            </svg>
          </div>
        </div>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CriteriaDropdown should render readOnly with selectedCriteriaKey 1`] = `
<DocumentFragment>
  <span
    class="bx--rule-criteria--readonly"
  />
</DocumentFragment>
`;
